ipsec: Attach to new ENI interfaces as they are added

pchaigno opened the pull request:
This pull request ensures that we rerun the detection and attachment of bpf_network to new ENI interfaces whenever they are added. See commits for details.

```release-note
Fix bug where Cilium wouldn't attach its IPsec BPF program to new ENI interfaces, resulting in connectivity loss between pods on remote nodes.  
```

pchaigno committed:
loader: Allow reinitializeIPSec to run multiple times

reinitializeIPSec only runs the interface detection if EncryptInterface
is empty. Since it sets it after detecting interfaces, it will only run
the detection once.

Let's change that to run the detection even if the EncryptInterface list
isn't empty. That will allow us to rerun the detection when new ENI
devices are added on EKS.

One consequence of this change is that we will now attach to all
interfaces even if the user configured --encrypt-interface. That is fine
because --encrypt-interface shouldn't actually be used in ENI mode. In
ENI mode, we want to attach to all interfaces as we don't have a
guarantee on which interface the IPsec traffic will come in.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
loader: Attach IPsec program to new ENI interfaces

We currently have a function reinitializeIPSec that detects any
interface likely to be an ENI interface and attaches bpf_network to it
to be able to decrypt IPsec traffic ingressing via that interface.
We however only run this function when the base datapath is reloaded
(Reinitialize). We don't currently reload the base datapath in reaction
to new ENI interfaces being added.

This commit changes that to ensure that, whenever we add a new ENI
interface, reinitializeIPSec is called and bpf_network attached to that
interface. To that end, if running IPsec in ENI mode, we subscribe the
datapath Loader to CiliumNode events. On update of CiliumNode objects,
we will check if any ENI was added and call reinitializeIPSec if that
is the case.

We can't simply call reinitializeIPSec when the new ENIs are added (vs.
when we see them propagated in CiliumNode) because the addition of ENI
interfaces happens in the operator and not in the agents.

reinitializeIPSec currently reloads bpf_network on all detected
interfaces, regardless of whether it is already attached or not. We
could probably optimize this to only attach to new interfaces, but given
this will only run when new ENIs are added, it probably doesn't matter
much. It probably also doesn't hurt to have this extra insurance that we
are always attached to all interfaces.

Fixes: 2c878a512de9 ("cilium: auto-detect network interfaces for IPAMENI case")
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno commented on pkg/datapath/loader/base.go:
```diff
@@ -499,3 +500,50 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+	// to them.
+	if len(newENIs) > 0 {
+		log.WithField(logfields.NewENIs, newENIs).Info("New ENIs detected")
+		return l.reinitializeIPSec(context.TODO())
```
For reviewers: How can I fix this to have a proper context?

joamaki commented on pkg/datapath/loader/base.go:
```diff
@@ -499,3 +500,50 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+// that case, we need to rerun reinitializeIPSec whenever new ENIs are added to
+// the local node, to attach our BPF program for IPsec decryption to the new
+// interfaces.
+func (l *Loader) OnUpdateCiliumNode(oldNode, newNode *ciliumv2.CiliumNode, _ *lock.StoppableWaitGroup) error {
```
I'd like to solve this in a cleaner way (and without k8s dependencies in pkg/datapath) by having loader observe the devices table (coming in https://github.com/cilium/cilium/pull/24677). I'm fine with this workaround until the proper ground work is done. Is there any pitfalls to just watching for the local network devices and addresses versus the CiliumNode object?

pchaigno replied:
> Is there any pitfalls to just watching for the local network devices and addresses versus the CiliumNode object?

I think that would work. Given Sebastian's comment below, it would probably even be easier once we have the new controller in place.

gandro reviewed (commented):
A high-level comment: We do have somewhat very similar code in https://github.com/cilium/cilium/blob/main/pkg/ipam/crd_eni.go 

One main difference seems to be that in `crd_eni` we wait for the netlink devices to be attached. This was needed, because sometimes the operator (which populates `.Status.ENI.ENIs` based on the state of the EC2 API) would know about ENIs before the Linux guest on the actual node saw it. I think similar logic is needed here, as having an ENI in `Status.ENI.ENIs` just means that AWS has attached the interface to the virtual machine, but not that the local Linux kernel has actually already discovered it and exposed it as a netlink interface.

As a side-note: This PR intended to be backported, so I think it's fine to have this fix architected as is for now. But long-term, I think we should probably introduce an "ENI device" manager (Jussi has suggested that in the past too) that merges the two pieces of code, since both seem to concern themselves with setting up the netlink interface.

gandro commented on pkg/datapath/loader/base.go:
```diff
@@ -499,3 +500,50 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+			}
+		}
+		if !existed {
+			newENIs = append(newENIs, newENI)
```
What interfaces do we actually want to attach the IPSec program to?

In IPAM, we actually do allow the user to exclude certain interfaces for pod IP allocation. This happens if the user wants to use those interfaces for other purposes than pod IPs, e.g. for egressgw or if they are running non-pod workloads on the node (see `spec.eni.exclude-interface-tags` and `spec.eni.first-interface-index` in https://docs.cilium.io/en/v1.13/network/concepts/ipam/eni/#ipam-eni)

Do we need to have the same logic here, or can the IPSec program be attached to any interface even if it doesn't have anything to do with Cilium?

Checking if an interface is excluded is luckily very easy:
https://github.com/cilium/cilium/blob/3e805306fc4125fca3398397acdf11a13f1a9600/pkg/ipam/crd_eni.go#L50-L52

pchaigno replied:
We can attach to any interface currently. That might change once we support IPsec+KPR, but right now, it's fine. We already attach to e.g. the `nodelocaldns` interface :shrug:  Unless those interface receive ESP traffic, the BPF program won't actually do anything.

gandro replied:
In that case I think it's also fine to just attach to those interfaces too. It's hard to judge how users treat the excluded interface, but we also never made any explicit guarantees, except that we're not using them for pod IP allocation.

schlosna commented on pkg/logging/logfields/logfields.go:
```diff
@@ -617,6 +617,9 @@ const (
 	ExpectedENIs = "expectedENIs"
 
+	// NewENIs are ENIs that have been added to the CiliumNode.
+	NewENIs = "newENIs"
```
`NewENIs` does not appear to be used

gandro commented on pkg/ipam/crd_eni.go:
```diff
@@ -104,6 +106,13 @@ func setupENIDevices(eniConfigByMac configMap) {
+
+	// Attach relevant BPF program to them.
+	if len(eniLinkByMac) > 0 {
+		if err := loader.ReinitializeIPsec(context.TODO()); err != nil {
```
There is one thing I just remembered which might be relevant: Because `setupENIDevices` runs in a go routine in the background, there might be multiple concurrent invocations of it.

So `ReinitializeIPsec` must be able to deal with the fact that it might be called concurrently. Which looking at the implementation, I don't think it is. Especially the synchronized write to `option.Config.EncryptInterface` looks like a potential data race to me.

(Another reason why we should refactor this whole logic here :disappointed: )


Also, feel free to introduce a context at the top of `setupENIDevices` like this:
```
func setupENIDevices(eniConfigByMac configMap) {
	ctx, cancel = context.WithCancel(context.Background())
	defer cancel()
```

Not sure why `setupENIDevices` did not use a context for e.g. the timeout, probably because it was implemented rather quickly to address a bug fix.

pchaigno committed:
eni: Refactor waitForNetlinkDevices to return missing ENIs

This commit refactors waitForNetlinkDevices to return the list of
missing ENI interfaces. This simplifies a bit the error handling for the
caller.

waitForNetlinkDevices also now sets the link (Linux interface) in the
ENI config to allow us to simply pass the list of ENI configs to other
functions.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
eni: Refactor configureENIDevices into separate parseENIConfigs

Splitting configureENIDevices will allow us to reuse the new
parseENIConfigs function in a subsequent commit.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
eni: Expose new function WaitForENIInterfaces

This function will parse new ENI devices from the given CiliumNodes and
wait for the corresponding Linux interfaces to show up before returning.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

gandro commented on pkg/datapath/loader/base.go:
```diff
@@ -511,3 +513,63 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+
+	// If we have any new ENIs, we first need to wait for the corresponding
+	// interfaces to show up...
+	missingENIByMac, err := ipam.WaitForENIInterfaces(oldNode, newNode)
```
This will block any CiliumNode events from being processed in the agent. We should move this into a go routine, a [job group](https://docs.cilium.io/en/latest/contributing/development/hive/#job-groups) or good old fashioned one-shot [controller](https://pkg.go.dev/github.com/cilium/cilium@v1.13.2/pkg/controller)

gandro commented on pkg/datapath/loader/base.go:
```diff
@@ -511,3 +513,63 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+		return err
+	}
+
+	return l.reinitializeIPSec(ctx)
```
Doesn't this need to be called mutually exclusive with `l.Reinitialize`? Otherwise both will write to shared global state without synchronization.
